templates/fine-tune-function-calling/README.md

```diff
@@ -0,0 +1,419 @@
+1. In case the ground truth response contains no function call, then the model's response should also not have a function call. 
+2. In case the ground truth response contains a function call, then the model's response should also have a function call. We do not check for the content of the response here. The assistant function call should further have the correct function name and the correct function arguments. 
+
+The following psuedocode shows some of the different branching conditions considered during evaluation:
+
+```
+if(ground_truth has no function call):
+    correct = (response has no function call)
+else
+    if response has no function call: 
+        correct = False
+    else
+          if response.function_name != gt.function_name:
+                correct = False
+          else
+                correct = (response.argument_dict == gt.argument_dict)
+```
+
+
+
+## Dataset formatting
+We follow the same preprocessing as during training for the finetuned model hosted on Anyscale. For GPT-4, the tool calls and tool responses are parsed and formatted in the OpenAI format. 
+
+We process the test dataset to have the following fields: 
+1. `openai_messages` : the list of messages in the conversation for GPT-4. The user messages will be fed to the model one by one. The function responses have to be designated with the role "tool" here.
+2. `anyscale_messages`: the list of messages in the conversation for our fine-tuned model.
+3. `tools`: the list of tools to pass to the OpenAI model. 
+4. `expected_responses` : the list of ground truth assistant responses in the conversation. 
+
+
+```python
+# preprocess the test dataset for evaluation
+modified_ds = get_evaluation_dataset(test_ds, TOOL_CALL_TAGS, TOOL_RESULT_TAGS)
+```
+
+## Evaluate
+
+For evaluation, we initialise two parsers - one for each model - to handle obtaining chat completions from the respective API and parsing the result. Then, our evaluation logic takes care of matching the assistant response with the expected response and, if the response is incorrect, making note of the type of error (wrong intent, wrong function name, etc). Populate the API keys below and run the below code blocks to get evaluation results:
+
+
+```python
+# enter your openai key below.
+OPENAI_API_KEY = "yourApiKeyHere" 
+# enter your Anyscale key below. If you finetuned through Anyscale endpoints, you can get the key here: https://console.anyscale.com/credentials. Otherwise, you should use the key from your Anyscale Service
+ANYSCALE_API_KEY = "yourApiKeyHere" 
+```
+
+
+```python
+# initialize parsers
+openai_parser = OpenAIResponseParser(api_key=OPENAI_API_KEY, api_base="https://api.openai.com/v1", model="gpt-4", tool_call_tags=TOOL_CALL_TAGS)
```

Comment:
"https://api.openai.com/v1" use a const definition above where you define the ANYSCALE_API_BASE?

Reply:
Have added this along with the API keys section above.